extensions/ql-vscode/src/skeleton-query-wizard.ts

```diff
@@ -0,0 +1,247 @@
+  python: "pallets/flask",
+  ruby: "rails/rails",
+  swift: "Alamofire/Alamofire",
+};
+
+export class SkeletonQueryWizard {
+  private language: string | undefined;
+  private fileName = "example.ql";
+  private storagePath: string | undefined;
+
+  constructor(
+    private readonly cliServer: CodeQLCliServer,
+    private readonly progress: ProgressCallback,
+    private readonly credentials: Credentials | undefined,
+    private readonly extLogger: OutputChannelLogger,
+    private readonly databaseManager: DatabaseManager,
+    private readonly token: CancellationToken,
+  ) {}
+
+  private get folderName() {
+    return `codeql-custom-queries-${this.language}`;
+  }
+
+  public async execute() {
+    // show quick pick to choose language
+    this.language = await this.chooseLanguage();
+    if (!this.language) {
+      return;
+    }
+
+    this.storagePath = this.getFirstStoragePath();
+
+    const skeletonPackAlreadyExists = isFolderAlreadyInWorkspace(
+      this.folderName,
+    );
+
+    if (skeletonPackAlreadyExists) {
+      // just create a new example query file in skeleton QL pack
+      await this.createExampleFile();
+      // select existing database for language
+      await this.selectExistingDatabase();
+    } else {
+      // generate a new skeleton QL pack with query file
+      await this.createQlPack();
+      // download database based on language and select it
+      await this.downloadDatabase();
+    }
+
+    // open a query file
+    await this.openExampleFile();
+  }
+
+  private async openExampleFile() {
+    if (this.folderName === undefined || this.storagePath === undefined) {
+      throw new Error("Path to folder is undefined");
+    }
+
+    const queryFileUri = Uri.file(
+      join(this.storagePath, this.folderName, this.fileName),
+    );
+
+    try {
+      void workspace.openTextDocument(queryFileUri).then((doc) => {
+        void Window.showTextDocument(doc);
+      });
+    } catch (e: unknown) {
+      void this.extLogger.log(
+        `Could not open example query file: ${getErrorMessage(e)}`,
+      );
+    }
+  }
+
+  public getFirstStoragePath() {
+    const workspaceFolders = workspace.workspaceFolders;
+
+    if (!workspaceFolders || workspaceFolders.length === 0) {
+      throw new Error("No workspace folders found");
+    }
+
+    const firstFolder = workspaceFolders[0];
+
+    // For the vscode-codeql-starter repo, the first folder will be a ql pack
+    // so we need to get the parent folder
+    if (firstFolder.uri.path.includes("codeql-custom-queries")) {
+      // slice off the last part of the path and return the parent folder
+      return firstFolder.uri.path.split("/").slice(0, -1).join("/");
+    } else {
+      // if the first folder is not a ql pack, then we are in a normal workspace
+      return firstFolder.uri.path;
+    }
+  }
+
+  private async chooseLanguage() {
+    this.progress({
+      message: "Choose language",
+      step: 1,
+      maxStep: 1,
```

Comment:
Why does this have a `maxStep` of 1? I would expect the `maxStep` to be consistent across all `this.progress` calls so that the user sees their progress increasing.

Reply:
Thanks 👍 I've opened a new PR to address this: https://github.com/github/vscode-codeql/pull/2305

I admit, I just copied it without really looking into how the steps were marked.